src/main/java/de/tum/cit/aet/artemis/programming/service/sshuserkeys/UserSshPublicKeyExpiryNotificationService.java

```diff
@@ -0,0 +1,57 @@
+package de.tum.cit.aet.artemis.programming.service.sshuserkeys;
+
+import static de.tum.cit.aet.artemis.core.config.Constants.PROFILE_SCHEDULING;
+
+import java.time.ZonedDateTime;
+
+import org.springframework.context.annotation.Profile;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Service;
+
+import de.tum.cit.aet.artemis.communication.service.notifications.SingleUserNotificationService;
+import de.tum.cit.aet.artemis.core.repository.UserRepository;
+import de.tum.cit.aet.artemis.programming.repository.UserSshPublicKeyRepository;
+
+@Profile(PROFILE_SCHEDULING)
+@Service
+public class UserSshPublicKeyExpiryNotificationService {
+
+    private final UserSshPublicKeyRepository userSshPublicKeyRepository;
+
+    private final SingleUserNotificationService singleUserNotificationService;
+
+    private final UserRepository userRepository;
+
+    public UserSshPublicKeyExpiryNotificationService(UserSshPublicKeyRepository userSshPublicKeyRepository, SingleUserNotificationService singleUserNotificationService,
+            UserRepository userRepository) {
+        this.userSshPublicKeyRepository = userSshPublicKeyRepository;
+        this.singleUserNotificationService = singleUserNotificationService;
+        this.userRepository = userRepository;
+    }
+
+    /**
+     * Notifies the user one week in advance about the upcoming expiry of one of their SSH keys
+     */
+    @Scheduled(cron = "0 0 7 * * *") // execute this every morning at 7:00:00 am
+    public void notifyUserOnUpcomingKeyExpiry() {
+        ZonedDateTime fromDate = ZonedDateTime.now().plusDays(6);
+        ZonedDateTime toDate = ZonedDateTime.now().plusDays(7);
+        userSshPublicKeyRepository.findByExpiryDateBetween(fromDate, toDate).forEach(userSshPublicKey -> {
+            var user = userRepository.findById(userSshPublicKey.getUserId());
+            user.ifPresent(recipient -> singleUserNotificationService.notifyUserAboutSoonExpiringSshKey(recipient, userSshPublicKey));
+        });
+    }
+
+    /**
+     * Notifies the user about the expiry of one of their SSH keys
+     */
+    @Scheduled(cron = "0 0 7 * * *") // execute this every morning at 7:00:00 am
+    public void notifyUserOnKeyExpiry() {
+        ZonedDateTime fromDate = ZonedDateTime.now().minusDays(1);
+        ZonedDateTime toDate = ZonedDateTime.now();
+        userSshPublicKeyRepository.findByExpiryDateBetween(fromDate, toDate).forEach(userSshPublicKey -> {
```

Comment:
can you load the users directly in one query? we should avoid loops with database requests if possible

Reply:
is there an index on `expiryDate` to speed up the query in case of many database entries?

Reply:
Yes, there is an index on expiryDate: https://github.com/ls1intum/Artemis/pull/9478/files#diff-1d90e0f0b2f992cdd43b2aef3b6d145862e99e55581f8c4bbed34059a15d173d

I also think loading all users at once here is better, so I changed it